Add `docvalue_fields` Support for `dense_vector` Fields

Rassyan opened the pull request:
### Summary

This PR implements support for the `docvalue_fields` parameter for `dense_vector` fields, as described in issue [#108470](https://github.com/elastic/elasticsearch/issues/108470).

### Details

- **Single Value Constraint**: 
  - Each document's `dense_vector` field contains a single vector value (multi-valued vectors are not allowed). The vector value is an array of numeric types. The `docvalue_fields` response will return this array directly as the value of `fields.vector`, without additional nesting.
  
- **Data Type Handling**:
  - For `byte` type `dense_vector`, the `docvalue_fields` response will be identical to the `_source` response.
  - For `float` type `dense_vector`, similar to other floating-point `docvalue_fields`, there may be minor precision differences between the `docvalue_fields` and `_source` responses. However, these differences are negligible and do not affect vector comparisons.

- **Implementation**:
  - Introduced a new `DenseDocValueFormat` class to handle `dense_vector` `docvalue_fields`. This class does not require actual formatting.
  - Overridden the `VectorDVLeafFieldData#getFormattedValues` method to return `FormattedDocValues` for `byte`, `bit`, and `float` type `dense_vector` fields.
  
- **Testing**:
  - Added YAML tests to cover the `docvalue_fields` parameter for `dense_vector` fields.

### Related Issues

Closes https://github.com/elastic/elasticsearch/issues/108470

Rassyan committed:
Enable docvalue_fields parameter for dense_vector fields

Rassyan committed:
Adding changelog for PR #114484

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

benwtrent committed:
Merge branch 'main' into desne_vector_docvalues_fields

mayya-sharipova commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/200_dense_vector_docvalue_fields.yml:
```diff
@@ -0,0 +1,111 @@
+  - length: {hits.hits.0.fields.vector5: 5}
+  - length: {hits.hits.0.fields.vector6: 5}
+
+  - lt: { hits.hits.0.fields.vector1.0: 230.01}
```
There is a new `close_to` assertion that can be used instead of `lt` and `gt`, as it

- close_to: { hits.hits.0.fields.vector1.0: { value: 230.0, error: 0.001 } }

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -99,6 +99,8 @@
  * A {@link FieldMapper} for indexing a dense vector of floats.
  */
 public class DenseVectorFieldMapper extends FieldMapper {
+    public static final NodeFeature DOCVALUES_FIELDS_SUPPORTED = new NodeFeature("search.vectors.docvalues_fields_supported");
```
let's group with other NodeFeature on line 112

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/DocValueFormat.java:
```diff
@@ -168,6 +168,31 @@ public String toString() {
+    /**
+     * Singleton, stateless formatter, for dense vector values, no need to actually format anything
+     */
+    class DenseDocValueFormat implements DocValueFormat {
```
Let's call this class `DenseVectorDocValueFormat`

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/DocValueFormat.java:
```diff
@@ -168,6 +168,31 @@ public String toString() {
         }
     };
 
+    DocValueFormat DENSE = DenseDocValueFormat.INSTANCE;
```
Let's call it `DENSE_VECTOR`

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -76,4 +87,109 @@ public DocValuesScriptFieldFactory getScriptFieldFactory(String name) {
+            public boolean advanceExact(int docID) throws IOException {
+                if (indexed) {
+                    final ByteVectorValues byteVectorValues = reader.getByteVectorValues(field);
+                    if (byteVectorValues.advance(docID) == NO_MORE_DOCS) {
```
This doesn't compile. Now we need to access vectors through iterator, as in [this example](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/script/field/vectors/ByteKnnDenseVectorDocValuesField.java#L50)

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/fielddata/DenseVectorNumericByteValues.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Use a simple array to store the values.
+ */
+public abstract class DenseVectorNumericByteValues {
```
Do we need the 2 abstract classes of `DenseVectorNumericByteValues` and `DenseVectorNumericFloatValues` ? Can we make implementation more concise and just introduce concrete classed in VectorDVLeafFieldData?

Rassyan replied:
You are right, the two abstract classes were not necessary. I have made the implementation more concise by introducing concrete classes directly in `VectorDVLeafFieldData`.

mayya-sharipova reviewed (commented):
@Rassyan Thanks for handling this issue.  Good start! I left some comments.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -76,4 +87,109 @@ public DocValuesScriptFieldFactory getScriptFieldFactory(String name) {
+            @Override
+            public boolean advanceExact(int docID) throws IOException {
+                if (indexed) {
+                    final FloatVectorValues floatVectorValues = reader.getFloatVectorValues(field);
```
You should not instantiate vector values on every iteration of docID.
They should be instantiated only once:

try {
            FloatVectorValues vectorValues = reader.getFloatVectorValues(field);

        } catch (IOException e) {
            throw new IllegalStateException("Cannot load doc values", e);
        }
and then use the save instance for iteration.

Rassyan replied:
Thanks, @mayya-sharipova! Oops, that was a rookie mistake. Fixed it now—only instantiating once and reusing. Appreciate the catch!

Rassyan committed:
Refactor code based on review

Rassyan commented:
Thank you, @mayya-sharipova , for your detailed feedback. I have made the following adjustments based on your suggestions:

1. Updated assertions to use `close_to` instead of `lt` and `gt`.
2. Removed `DOCVALUES_FIELDS_SUPPORTED` NodeFeature.
3. Renamed the class to `DenseVectorDocValueFormat`.
4. Renamed the constant to `DENSE_VECTOR`.
5. Adjusted the code to access vectors through an iterator as per the provided example.
6. Simplified the implementation by removing the two abstract classes `DenseVectorNumericByteValues` and `DenseVectorNumericFloatValues`, and introduced concrete classes in `VectorDVLeafFieldData`.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/fielddata/plain/DenseVectorNumericValues.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Override
+    public final int docValueCount() {
```
`docValueCount` means how many values of this Object we have.  For vectors all of them are single valued. This should return `1` if a document has a vector value

Rassyan replied:
Let me explain the logic behind the single value constraint we mentioned earlier:

> Each document's `dense_vector` field contains a single vector value, meaning multi-valued vectors are not allowed. The vector value is an array of numeric types. The `docvalue_fields` response will return this array directly as `fields.vector`, without additional nesting.

This logic is reflected here. If `docValueCount` returns 1, the result in `hits.hits.0.fields.vector` would be `[[1,2,3]]`. By returning the dimensions (`dims`), the result in `hits.hits.0.fields.vector` will be `[1,2,3]`. Since all vectors are single-valued, the former format is redundant, and the latter is more elegant. Additionally, when specifying `fields` parameter in the DSL, the return format is also `[1,2,3]`. This approach unifies the responses of `docvalue_fields` and `fields`.

mayya-sharipova replied:
I understand your reasons, it is indeed nice to have the output of `docvalue_fields` and `fields` to be the same, but I would not worry about this, we have differences for other field types as well. 

`docvalue_fields` have a predefined interface. `doValueCount` must return `1` as this is the contract we have: ` A count of the number of values at this document.`, and it should be 1. Also `docvalue_fields` always return an array of values, for a vector field a value is an array, so we need to return an array of arrays.

So, let's do the following:
- return 1 
- remove valueCursor

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -76,4 +86,94 @@ public DocValuesScriptFieldFactory getScriptFieldFactory(String name) {
+
+                @Override
+                public Object nextValue() {
+                    return values[valuesCursor++];
```
We don't need `valueCursor`. I **think** `nextValue` can always return the same value

```java
public Object nextValue() throws IOException {
      return values;
}
```

Rassyan replied:
Same issue as `docValueCount` returning 1.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/fielddata/plain/DenseVectorNumericValues.java:
```diff
@@ -0,0 +1,41 @@
+import org.apache.lucene.index.KnnVectorValues;
+import org.elasticsearch.index.fielddata.FormattedDocValues;
+
+public abstract class DenseVectorNumericValues implements FormattedDocValues {
```
Do we need this class at all? Can we  just use `FormattedDocValues` in `VectorDVLeafFieldData`?

Rassyan replied:
No, we can use `FormattedDocValues`.

mayya-sharipova reviewed (commented):
@Rassyan Thanks for iterating. The change looks much better now. I've a couple more comments.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -76,4 +86,94 @@ public DocValuesScriptFieldFactory getScriptFieldFactory(String name) {
+                @Override
+                public boolean advanceExact(int docId) throws IOException {
+                    if (indexed) {
+                        if (floatVectorValues == null) {
```
can we have 1 `if` statement: 

```java
if (floatVectorValues == null) {
   floatVectorValues = reader.getFloatVectorValues(field);
    iterator = floatVectorValues.iterator();
}
```
iterator is coupled with floatVectorValues and you must modify it when floatVectorvalues change.

Rassyan replied:
Understood, will do.

Rassyan committed:
Refactor code based on review

Rassyan commented:
@mayya-sharipova, thank you for your comments. I've already incorporated some of your suggestions and am looking forward to more of your feedback.

Rassyan committed:
Refactor code based on review

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -76,4 +88,109 @@ public DocValuesScriptFieldFactory getScriptFieldFactory(String name) {
+                            this.floatVectorValues = reader.getFloatVectorValues(field);
+                            this.iterator = floatVectorValues.iterator();
+                        }
+                        if (iterator.advance(docId) == NO_MORE_DOCS) {
+                            return false;
+                        }
+                        values = new ArrayList<>(dims);
+                        for (float f : floatVectorValues.vectorValue(iterator.index())) {
+                            values.add(f);
+                        }
+                    } else {
+                        if (binary == null) {
+                            binary = DocValues.getBinary(reader, field);
+                        }
+                        if (binary.advance(docId) == NO_MORE_DOCS) {
+                            return false;
+                        }
+                        BytesRef ref = binary.binaryValue();
+                        ByteBuffer byteBuffer = ByteBuffer.wrap(ref.bytes, ref.offset, ref.length);
+                        if (indexVersion.onOrAfter(DenseVectorFieldMapper.LITTLE_ENDIAN_FLOAT_STORED_INDEX_VERSION)) {
+                            byteBuffer.order(ByteOrder.LITTLE_ENDIAN);
+                        }
+                        values = new ArrayList<>(dims);
+                        for (int dim = 0; dim < dims; dim++) {
+                            values.add(byteBuffer.getFloat());
+                        }
+                    }
+                    return true;
+                }
+
+                @Override
+                public int docValueCount() {
+                    return 1;
+                }
+
+                @Override
+                public Object nextValue() {
+                    return values;
+                }
+            };
```
```suggestion
            case FLOAT -> new FormattedDocValues() {
                float[] vector = new float[dims];
                private FloatVectorValues floatVectorValues; // use when indexed
                private KnnVectorValues.DocIndexIterator iterator; // use when indexed
                private BinaryDocValues binary; // use when not indexed

                @Override
                public boolean advanceExact(int docId) throws IOException {
                    if (indexed) {
                        if (floatVectorValues == null) {
                            this.floatVectorValues = reader.getFloatVectorValues(field);
                            this.iterator = floatVectorValues.iterator();
                        }
                        if (iterator.advance(docId) == NO_MORE_DOCS) {
                            return false;
                        }
                        vector = floatVectorValues.vectorValue(iterator.index());
                    } else {
                        if (binary == null) {
                            binary = DocValues.getBinary(reader, field);
                        }
                        if (binary.advance(docId) == NO_MORE_DOCS) {
                            return false;
                        }
                        BytesRef ref = binary.binaryValue();
                        VectorEncoderDecoder.decodeDenseVector(indexVersion, ref, vector);
                    }
                    return true;
                }

                @Override
                public int docValueCount() {
                    return 1;
                }

                @Override
                public Object nextValue() {
                    return vector;
                }
            };
```

mayya-sharipova replied:
@Rassyan please do similar changes for `BYTE, BIT` part, in particular:
- use utility function to  decode vector as in`VectorEncoderDecoder.decodeDenseVector`as in the future we may change the format
- don't create objects like: `values = new ArrayList<>(dims);` on every doc iteration.

Rassyan replied:
@mayya-sharipova, understood. For the byte type, I couldn't find an existing implementation, so I will refer to `VectorEncoderDecoder.decodeDenseVector` to implement byte vector decoding.

Regarding the choice between `byte[]` and `List<Byte>`, I opted for `ArrayList` for the following reasons:

1. In the final response, we expect the output to be like `{"vector5":[[1,111,-13,15,-128]],"vector6":[[-1,11,0,12,111]]}`. If we return a `byte[]` in `nextValue()`, it gets serialized to a string in the response: `{"vector5":["AW/zD4A="],"vector6":["/wsADG8="]}`.
2. In YAML tests, assertions for match and length only accept `List` types and cannot handle array types.

Given these points, I chose to use `List<Byte>`. Let's discuss a suitable solution further. Looking forward to your feedback.

mayya-sharipova replied:
I understood, thanks for explanation. Let me think about it and will get back to you

mayya-sharipova replied:
@Rassyan I pushed the correction as the last commit: bf64fc3ac3553c9e69cc39b2960fbe44b993a443. Please check it out. Two things addressed:

- You were right, we can not return just byte[] array. I returned array of object Byte
- for iteration over knn values we need to implement advaceExact, which I added.

Rassyan replied:
I've reviewed commit bf64fc3. Your corrections are spot on, and I appreciate your hands-on guidance. Thanks for the detailed attention!

mayya-sharipova committed:
Correction

mayya-sharipova commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/200_dense_vector_docvalue_fields.yml:
```diff
@@ -0,0 +1,118 @@
+        body:
+          name: moose.jpg
+
+  - do:
```
Let's add a couple of more documents after that with some vector fields missing and some present to ensure correct iteration.

Rassyan replied:
Got it, I've added the additional documents to test the iteration.

mayya-sharipova commented:
@Rassyan I think the solution I provided in the last commit should work.

Let's make the test more advanced, and I think this PR will be ready after that.

Rassyan committed:
Advanced test

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
@Rassyan I started CI, and some tests are failing: 
```
REPRODUCE WITH: ./gradlew ":server:test" --tests "org.elasticsearch.index.mapper.vectors.DenseVectorFieldTypeTests.testDocValueFormat" -Dtests.seed=5E3E0C56725338D7 -Dtests.locale=ebu-KE -Dtests.timezone=Australia/West -Druntime.java=22
  |  
  | DenseVectorFieldTypeTests > testDocValueFormat FAILED
  | junit.framework.AssertionFailedError: Expected exception IllegalArgumentException but no exception was thrown
  | at __randomizedtesting.SeedInfo.seed([5E3E0C56725338D7:ACD632D4793932D0]:0)
  | at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2889)
  | at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2875)
  | at org.elasticsearch.index.mapper.vectors.DenseVectorFieldTypeTests.testDocValueFormat(DenseVectorFieldTypeTests.java:137)
 

<br class="Apple-interchange-newline">
```

Please fix them.

Rassyan committed:
Fix DenseVectorFieldTypeTests.testDocValueFormat

mayya-sharipova commented:
@elasticsearchmachine test this please

mayya-sharipova commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into desne_vector_docvalues_fields

mayya-sharipova commented:
@elastimachine test this please

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into desne_vector_docvalues_fields

mayya-sharipova committed:
Mode edits

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/part-4

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/windows-2019 / default-windows-archive / packaging-tests-windows-sample

mayya-sharipova commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into desne_vector_docvalues_fields

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/bwc-snapshots

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into desne_vector_docvalues_fields

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
@elasticmachine run "Elasticsearch Serverless Checks"

mayya-sharipova committed:
Ensure consistent sorting of docs

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
@Rassyan Thank you for your work and congratulations, this has been merged for 8.17 release.

benwtrent commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/116491">](https://github.com/elastic/elasticsearch/pull/116491)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.1 -->
